### PR TITLE
HS-1484: Consolidate tag creation at Inventory service

### PR DESCRIPTION
**Pull request by cgorantla**
## Jira link(s)
- https://opennms.atlassian.net/browse/HS-1484

## Checklist
* [ ] Follows Lōkahi's [development guidelines.](https://github.com/OpenNMS-Cloud/lokahi/wiki/Development-Guidelines)
* [ ] Appropriate reviewer(s) have been selected.
* [ ] Jira issue(s) have been updated to "In Review".
* [ ] Includes [appropriate tests.](https://github.com/OpenNMS-Cloud/lokahi/wiki/Test-Strategy)
* [ ] Documentation has been updated as necessary.
* [ ] Notify devops of changes to the Charts

### CODE REVIEW

File: inventory/main/src/main/java/org/opennms/horizon/inventory/model/Tag.java [hunk 1]
```diff
@@ -85,4 +85,11 @@ public class Tag {
         joinColumns = @JoinColumn(name = "tag_id"),
         inverseJoinColumns = @JoinColumn(name = "passive_discovery_id"))
     private List<PassiveDiscovery> passiveDiscoveries = new ArrayList<>();
+
+/*    @ManyToMany
```
Comment (fooker) [inline inventory/main/src/main/java/org/opennms/horizon/inventory/model/Tag.java:89]: Any reason why this is skipped?

Comment (cgorantla) [reply inventory/main/src/main/java/org/opennms/horizon/inventory/model/Tag.java]: MonitoringPolicy is only defined in Alarm service, here we are only  persisting monitoring policy id with a separate table `monitoring_policy_tag`

**Review by splatch — COMMENTED** [review]
Looks good, some small remarks on things I found possible to improve.

File: rest-server/src/main/java/org/opennms/horizon/server/service/GrpcAlertService.java [hunk 2]
```diff
@@ -109,7 +121,10 @@ public Mono<DeleteAlertResponse> deleteAlert(@GraphQLArgument(name = "ids") List
 
     @GraphQLMutation
     public Mono<MonitorPolicy> createMonitorPolicy(MonitorPolicy policy, @GraphQLEnvironment ResolutionEnvironment env) {
-        return Mono.just(alertsClient.createMonitorPolicy(policy, headerUtil.getAuthHeader(env)));
+        String authHeader = headerUtil.getAuthHeader(env);
+        var monitorPolicy = alertsClient.createMonitorPolicy(policy, headerUtil.getAuthHeader(env));
+        createTagsInInventory(authHeader, monitorPolicy.getId(), policy.getTags());
+        return Mono.just(monitorPolicy);
```
[the remark this answers is not in the document]
Comment (cgorantla) [reply rest-server/src/main/java/org/opennms/horizon/server/service/GrpcAlertService.java:128]: I think we need to use result of first in 2nd, so not sure if we can still use Mono.zip with that.
I'm still rusty on reactor bits, will check this later.

File: inventory/main/src/test/java/org/opennms/horizon/inventory/repository/TagServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.test.context.ContextConfiguration;
+
+@SpringBootTest
+@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE)
+@ContextConfiguration(initializers = {SpringContextTestInitializer.class})
+@AutoConfigureObservability
+// Make sure to include Metrics (for some reason they are disabled by default in the integration grey-box test)
+public class TagServiceTest {
```
[the remark this answers is not in the document]
Comment (cgorantla) [reply inventory/main/src/test/java/org/opennms/horizon/inventory/repository/TagServiceTest.java:52]: Retained this but disabled ( need to keep these developer tests)

File: inventory/main/src/main/java/org/opennms/horizon/inventory/model/Tag.java [hunk 4]
```diff
@@ -85,4 +86,8 @@ public class Tag {
         joinColumns = @JoinColumn(name = "tag_id"),
         inverseJoinColumns = @JoinColumn(name = "passive_discovery_id"))
     private List<PassiveDiscovery> passiveDiscoveries = new ArrayList<>();
+
+    @Column(name = "policy_ids", columnDefinition = "jsonb")
```
[the remark this answers is not in the document]
Comment (cgorantla) [reply inventory/main/src/main/java/org/opennms/horizon/inventory/model/Tag.java]: Handled

File: alert/pom.xml [hunk 5]
```diff
@@ -234,6 +234,12 @@
             <artifactId>java-jwt</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>com.h2database</groupId>
+            <artifactId>h2</artifactId>
+            <version>2.1.214</version>
```
Comment (freddy61025) [inline alert/pom.xml:240]: suggest to put version in parent-pom

**Review by freddy61025 — APPROVED** [review]
Some minor comments. Please make sure database change will not cause problem for existing data.

File: inventory/main/src/main/resources/db/changelog/hs-0.1.0/tables/tag.xml [hunk 6]
```diff
@@ -44,4 +46,11 @@
                                  referencedTableName="tag" referencedColumnNames="id"/>
 
     </changeSet>
+
+    <changeSet id="0.1.0-table-add-policy-ids" author="cgorantla">
+        <addColumn tableName="tag">
+            <column name="monitor_policy_ids" type="jsonb" />
```
[the remark this answers is not in the document]
Comment (cgorantla) [reply inventory/main/src/main/resources/db/changelog/hs-0.1.0/tables/tag.xml:52]: not sure if I understand this correctly.  This is a new column that we are adding.  Monitor policies are being created in Alert service and we will not persist if there is a monitoring policy with invalid id.

**Review by splatch — APPROVED** [review]
Looks good, thank you for updating column names.

I am a bit afraid  of actual approach. As far I can see we do rely on two backend calls made for a single UI request. I do see a risk of running out of sync, however that's something we can address in further updates to this functionality.

**Comment by cgorantla** [comment]
> Looks good, thank you for updating column names.
> 
> I am a bit afraid of actual approach. As far I can see we do rely on two backend calls made for a single UI request. I do see a risk of running out of sync, however that's something we can address in further updates to this functionality.

I think it's better to sync tags from one service to other by subscribing to messages instead of updating multiple services.
Will be handled in https://opennms.atlassian.net/browse/HS-1769

**Pull request merged**